app/workers/push_notifications/cleanup_worker.rb

```diff
@@ -0,0 +1,18 @@
+module PushNotifications
+  class CleanupWorker
+    include Sidekiq::Worker
+
+    sidekiq_options queue: :low_priority, retry: 10
+
+    def perform
+      redis = Redis.new
+      redis.keys("rpush:notifications:*").each do |key|
```

Comment:
`KEYS` is an O(n) operation in Redis, which we should really avoid if we can. It will block not just this worker, but it puts every client of this Redis instance in a scenario similar to [head-of-line blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking). Nobody will be able to talk to it at all until it finishes.

A preferred operation would be using [`SCAN`](https://redis.io/commands/scan):

```ruby
cursor = 0
until (cursor, keys = redis.scan(cursor)).first.to_i == 0
  keys.each do |key|
    # ...
  end
end
```

My `until` is unnecessarily clever, but maybe we could refactor it to something more legible.

This is definitely a slower way to do it, but blocking Redis for an unknown amount of time sounds scary.

Reply:
What @jgaskins said. FWIW, the `#scan` method in the Ruby gem can also use [`SSCAN`](https://redis.io/commands/SSCAN), e.g. `scan(cursor, match: "rpush:notifications:*")`.

Reply:
This snippet was taken as-is from their [Redis database cleanup section](https://github.com/rpush/rpush/wiki/Using-Redis#database-cleanup), so using the alternatives proposed by both @jgaskins and @citizen428 sound great to me. The end goal is the same (ensure all rpush notifications will eventually expire) but this would make it more efficient 👍

Reply:
@citizen428 Oh, right, I didn't put the `match` in there. Nice catch!

Was `SSCAN` a typo, though? I thought this was going over top-level keys rather than items in a set.

Reply:
Not quite a typo, more like me being on the wrong page and then confusing myself. :D

Reply:
✅